File: CTRebuildAndroid/app/src/main/java/com/chasetactical/ctrebuild/models/ColumnDefinition.kt
package com.chasetactical.ctrebuild.models

import org.json.JSONArray
import org.json.JSONObject

enum class DataKind {
    Text, Number, Dropdown, Date, BarcodeScan, Photo, Gps, ManualEntry, Toggle, Computed;

    companion object {
        fun fromString(value: String?): DataKind =
            entries.firstOrNull { it.name.equals(value, ignoreCase = true) } ?: Text
    }
}

data class ColumnDefinition(
    val id: String,
    val tableName: String,
    val headerText: String,
    val dataKind: DataKind,
    val sortOrder: Int,
    val createdAtUtc: String,
    val options: List<String>?,
    val bindingPath: String?,
    val isReadOnly: Boolean,
    val defaultWidth: Double,
    val isCollapsibleId: Boolean
) {
    companion object {
        fun fromJson(obj: JSONObject): ColumnDefinition {
            val optionsArray = obj.optJSONArray("options")
            val options = optionsArray?.let {
                List(it.length()) { i -> it.getString(i) }
            }
            return ColumnDefinition(
                id             = obj.getString("id"),
                tableName      = obj.optString("tableName", ""),
                headerText     = obj.optString("headerText", ""),
                dataKind       = DataKind.fromString(obj.optString("dataKind")),
                sortOrder      = obj.optInt("sortOrder", 0),
                createdAtUtc   = obj.optString("createdAtUtc", ""),
                options        = options,
                bindingPath    = obj.optString("bindingPath").ifBlank { null },
                isReadOnly     = obj.optBoolean("isReadOnly", false),
                defaultWidth   = obj.optDouble("defaultWidth", 0.0),
                isCollapsibleId = obj.optBoolean("isCollapsibleId", false)
            )
        }

        fun listFromJsonArray(array: JSONArray): List<ColumnDefinition> =
            List(array.length()) { i -> fromJson(array.getJSONObject(i)) }
    }

    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id)
        put("tableName", tableName)
        put("headerText", headerText)
        put("dataKind", dataKind.name)
        put("sortOrder", sortOrder)
        put("createdAtUtc", createdAtUtc)
        if (options != null) put("options", JSONArray(options))
        if (bindingPath != null) put("bindingPath", bindingPath)
        put("isReadOnly", isReadOnly)
        put("defaultWidth", defaultWidth)
        put("isCollapsibleId", isCollapsibleId)
    }
}

File: CT-Hub/Models/ColumnDefinition.cs
using System.Text.Json.Serialization;

namespace CTHub.Models;

public enum DataKind
{
    Text,
    Number,
    Dropdown,
    Date,
    BarcodeScan,
    Photo,
    Gps,
    ManualEntry,
    Toggle,
    Computed
}

public class ColumnDefinition
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [JsonPropertyName("tableName")]
    public string TableName { get; set; } = string.Empty;

    [JsonPropertyName("headerText")]
    public string HeaderText { get; set; } = string.Empty;

    [JsonPropertyName("dataKind")]
    public DataKind DataKind { get; set; } = DataKind.Text;

    [JsonPropertyName("sortOrder")]
    public int SortOrder { get; set; } = 0;

    [JsonPropertyName("createdAtUtc")]
    public string CreatedAtUtc { get; set; } = DateTime.UtcNow.ToString("o");

    /// <summary>
    /// Selectable values when DataKind is Dropdown.
    /// </summary>
    [JsonPropertyName("options")]
    public List<string>? Options { get; set; }

    /// <summary>
    /// Typed model property path (e.g. "Bin", "Qty") for seeded columns.
    /// Null/empty = extra field, bound via indexer [Id].
    /// </summary>
    [JsonPropertyName("bindingPath")]
    public string? BindingPath { get; set; }

    /// <summary>If true the column is read-only in the grid.</summary>
    [JsonPropertyName("isReadOnly")]
    public bool IsReadOnly { get; set; }

    /// <summary>
    /// Preferred column width in pixels. 0 = default (140 px). -1 = star (fill remaining).
    /// </summary>
    [JsonPropertyName("defaultWidth")]
    public double DefaultWidth { get; set; }

    /// <summary>If true the column is the collapsible ID column (grey, small font, pinned last).</summary>
    [JsonPropertyName("isCollapsibleId")]
    public bool IsCollapsibleId { get; set; }

    /// <summary>
    /// Signed column-ID formula for Computed columns. Format: "+colId1 +colId2 -colId3".
    /// </summary>
    [JsonPropertyName("formula")]
    public string? Formula { get; set; }

    /// <summary>Column-level warning threshold for Number columns. Cells at/below this value are highlighted red; within 25% above are highlighted amber.</summary>
    [JsonPropertyName("warningThreshold")]
    public double? WarningThreshold { get; set; }
}
